api/app.py

```diff
@@ -0,0 +1,354 @@
+    """
+    Upload the results to the webhook.
+    """
+
+    log(f"Uploading results to webhook: {data_dict['callback_url']}")
+    if data_dict["callback_method"] != "POST":
+        log("Callback URL is not a POST URL. Skipping.")
+        return
+    url = data_dict["callback_url"]
+    try:
+        _ = requests.post(
+            url,
+            json=results,
+            timeout=request_timeout,
+        )
+    except requests.exceptions.Timeout:
+        # If we timeout here then it's precisely because
+        # we cannot write back to the SkyPortal instance.
+        # So returning something does't make sense in this case.
+        # Just print it and move on...
+        log("Callback URL timedout. Skipping.")
+    except Exception as e:
+        log(f"Callback exception {e}.")
+
+
+def run_nmma_model(data_dict):
+    """
+    Use `nmma` to fit data to a model with name `source_name`.
+
+    For this analysis, we expect the `inputs` dictionary to have the following keys:
+       - source: the name of the model to fit to the data
+       - fix_z: whether to fix the redshift
+       - photometry: the photometry to fit to the model (in csv format)
+       - redshift: the known redshift of the object
+
+    Other analysis services may require additional keys in the `inputs` dictionary.
+    """
+    analysis_parameters = data_dict["inputs"].get("analysis_parameters", {})
+    analysis_parameters = {**default_analysis_parameters, **analysis_parameters}
+
+    source = analysis_parameters.get("source")
+    fix_z = analysis_parameters.get("fix_z") in [True, "True", "t", "true"]
+
+    # this example analysis service expects the photometry to be in
+    # a csv file (at data_dict["inputs"]["photometry"]) with the following columns
+    # - filter: the name of the bandpass
+    # - mjd: the modified Julian date of the observation
+    # - magsys: the mag system (e.g. ab) of the observations
+    # - flux: the flux of the observation
+    #
+    # the following code transforms these inputs from SkyPortal
+    # to the format expected by nmma.
+    #
+    rez = {"status": "failure", "message": "", "analysis": {}}
+    try:
+        data = Table.read(data_dict["inputs"]["photometry"], format="ascii.csv")
+        redshift = Table.read(data_dict["inputs"]["redshift"], format="ascii.csv")
+        z = redshift["redshift"][0]
+    except Exception as e:
+        rez.update(
+            {
+                "status": "failure",
+                "message": f"input data is not in the expected format {e}",
+            }
+        )
+        return rez
+
+    # we will need to write to temp files
+    # locally and then write their contents
+    # to the results dictionary for uploading
+    local_temp_files = []
+    cand_name = data_dict["resource_id"]
+
+    prior_directory = f"{os.path.dirname(os.path.realpath(__file__))}/priors"
+    svdmodel_directory = f"{os.path.dirname(os.path.realpath(__file__))}/svdmodels"
```

Comment:
It looks like the Dockerfile will take care of this, but for local testing I needed to copy the `priors` and `svdmodels` directories to within the `api` directory.